In [1]:
# Import required libraries

import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

! conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

Solving environment: done


==> WARNING: A newer version of conda exists. <==
  current version: 4.5.11
  latest version: 4.8.2

Please update conda by running

    $ conda update -n base -c defaults conda



## Package Plan ##

  environment location: /home/jupyterlab/conda/envs/python

  added / updated specs: 
    - geopy


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    certifi-2019.11.28         |           py36_0         149 KB  conda-forge
    scikit-learn-0.20.1        |   py36h22eb022_0         5.7 MB
    liblapack-3.8.0            |      11_openblas          10 KB  conda-forge
    numpy-1.18.1               |   py36h95a1406_0         5.2 MB  conda-forge
    liblapacke-3.8.0           |      11_openblas          10 KB  conda-forge
    geographiclib-1.50         |             py_0          34 KB  conda-forge
    libopenblas-0.3.6          |       h5a2b251_2         7.7 MB
    sci

In [2]:
# install and import a library to scrape the Wikipedia page 

! pip install beautifulsoup4

     |████████████████████████████████| 112kB 7.6MB/s eta 0:00:01


In [3]:
from bs4 import BeautifulSoup

In [4]:
# scrape the page
url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
page = requests.get(url)

In [5]:
soup = BeautifulSoup(page.content, 'html.parser')
tb = soup.find('table', class_='wikitable')
tb
all_links = tb.find_all("td")

In [6]:
all_links


[<td>M1A</td>, <td>Not assigned</td>, <td>Not assigned
 </td>, <td>M2A</td>, <td>Not assigned</td>, <td>Not assigned
 </td>, <td>M3A</td>, <td><a href="/wiki/North_York" title="North York">North York</a></td>, <td><a href="/wiki/Parkwoods" title="Parkwoods">Parkwoods</a>
 </td>, <td>M4A</td>, <td><a href="/wiki/North_York" title="North York">North York</a></td>, <td><a href="/wiki/Victoria_Village" title="Victoria Village">Victoria Village</a>
 </td>, <td>M5A</td>, <td><a href="/wiki/Downtown_Toronto" title="Downtown Toronto">Downtown Toronto</a></td>, <td><a href="/wiki/Regent_Park" title="Regent Park">Harbourfront</a>
 </td>, <td>M6A</td>, <td><a href="/wiki/North_York" title="North York">North York</a></td>, <td><a href="/wiki/Lawrence_Heights" title="Lawrence Heights">Lawrence Heights</a>
 </td>, <td>M6A</td>, <td><a href="/wiki/North_York" title="North York">North York</a></td>, <td><a href="/wiki/Lawrence_Manor" title="Lawrence Manor">Lawrence Manor</a>
 </td>, <td>M7A</td>, <td>

In [7]:
# extract titles and put in list
flist =[]
for link in all_links:
    name = link.get_text('title').strip()
    if name[-5:]=='title': name = name[:-5]
    flist.append(name)

In [8]:
# reshape data
np.asarray(flist)
flist=np.reshape(flist,(-1,3))


In [9]:
# merge neighbourhoods with the same postal code
pc = {}
for i in range(flist.shape[0]):
    pc.setdefault(flist[i,0],['',''])[0]=flist[i,1]
    pc[flist[i,0]][1]=pc[flist[i,0]][1]+', '+ flist[i,2]
    pc[flist[i,0]][1]= pc[flist[i,0]][1].strip(', ')
pc_tor = []    
for i in pc.keys():
    pc_tor.append([i, pc[i][0],pc[i][1]])

In [10]:
# put data into Data Frame and drop Not assigned
toronto = pd.DataFrame(pc_tor, columns=['Postal Code', 'Borough', 'Neighborhood'])
toronto.replace('Not assigned', np.nan, inplace = True)
toronto.dropna(axis=0, inplace = True)
toronto.reset_index(inplace=True, drop = True)

In [11]:
# check the result
toronto

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
5           M9A         Etobicoke   
6           M1B       Scarborough   
7           M3B        North York   
8           M4B         East York   
9           M5B  Downtown Toronto   
10          M6B        North York   
11          M9B         Etobicoke   
12          M1C       Scarborough   
13          M3C        North York   
14          M4C         East York   
15          M5C  Downtown Toronto   
16          M6C              York   
17          M9C         Etobicoke   
18          M1E       Scarborough   
19          M4E      East Toronto   
20          M5E  Downtown Toronto   
21          M6E              York   
22          M1G       Scarborough   
23          M4G         East York   
24          M5G  Downtown Toronto   
25          M6G  Downtown Toronto   
26          M1H       Scarborough   
27          M2H        North York   
28          M3H        North York   
29          M4H         East York   
30          M5H  Downtown Toronto   
31          M6H      West Toronto   
32          M1J       Scarborough   
33          M2J        North York   
34          M3J        North York   
35          M4J         East York   
36          M5J  Downtown Toronto   
37          M6J      West Toronto   
38          M1K       Scarborough   
39          M2K        North York   
40          M3K        North York   
41          M4K      East Toronto   
42          M5K  Downtown Toronto   
43          M6K      West Toronto   
44          M1L       Scarborough   
45          M2L        North York   
46          M3L        North York   
47          M4L      East Toronto   
48          M5L  Downtown Toronto   
49          M6L        North York   
50          M9L        North York   
51          M1M       Scarborough   
52          M2M        North York   
53          M3M        North York   
54          M4M      East Toronto   
55          M5M        North York   
56          M6M              York   
57          M9M        North York   
58          M1N       Scarborough   
59          M2N        North York   
60          M3N        North York   
61          M4N   Central Toronto   
62          M5N   Central Toronto   
63          M6N              York   
64          M9N              York   
65          M1P       Scarborough   
66          M2P        North York   
67          M4P   Central Toronto   
68          M5P   Central Toronto   
69          M6P      West Toronto   
70          M9P         Etobicoke   
71          M1R       Scarborough   
72          M2R        North York   
73          M4R   Central Toronto   
74          M5R   Central Toronto   
75          M6R      West Toronto   
76          M7R       Mississauga   
77          M9R         Etobicoke   
78          M1S       Scarborough   
79          M4S   Central Toronto   
80          M5S  Downtown Toronto   
81          M6S      West Toronto   
82          M1T       Scarborough   
83          M4T   Central Toronto   
84          M5T  Downtown Toronto   
85          M1V       Scarborough   
86          M4V   Central Toronto   
87          M5V  Downtown Toronto   
88          M8V         Etobicoke   
89          M9V         Etobicoke   
90          M1W       Scarborough   
91          M4W  Downtown Toronto   
92          M5W  Downtown Toronto   
93          M8W         Etobicoke   
94          M9W         Etobicoke   
95          M1X       Scarborough   
96          M4X  Downtown Toronto   
97          M5X  Downtown Toronto   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                          Neighborhood  
0                                            Parkwoods  
1                                     Vic

In [12]:
toronto.shape

(103, 3)

# The end of the first part

In [13]:
# install Geocoder
! pip install geocoder

     |████████████████████████████████| 102kB 7.7MB/s ta 0:00:011
     |████████████████████████████████| 829kB 12.6MB/s eta 0:00:01
     |████████████████████████████████| 81kB 15.8MB/s eta 0:00:01
  Stored in directory: /home/jupyterlab/.cache/pip/wheels/8b/99/a0/81daf51dcd359a9377b110a8a886b3895921802d2fc1b2397e
Successfully built future


import geocoder

latitude=[]
longitude=[]
for i in range(toronto.shape[0]):
    lat_lng_coords = None
    print(i)
    PCode = toronto.iloc[i]['Postal Code']+ ', Toronto, Ontario'
    print(PCode)
    while(lat_lng_coords is None):
        g = geocoder.google(PCode)
        lat_lng_coords = g.latlng
    latitude.append(lat_lng_coords[0])
    longitude.append(lat_lng_coords[1])
    

## tried many times and ways to get latitude and longitude, but failed. So decided to download geo data

In [14]:
geo_data = pd.read_csv('Geospatial_Coordinates.csv')

In [16]:
geo_data.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [17]:
toronto_data = pd.merge(toronto, geo_data, on='Postal Code')

In [18]:
toronto_data.head(10)

Postal Code           Borough                      Neighborhood   Latitude  \
0         M3A        North York                         Parkwoods  43.753259   
1         M4A        North York                  Victoria Village  43.725882   
2         M5A  Downtown Toronto                      Harbourfront  43.654260   
3         M6A        North York  Lawrence Heights, Lawrence Manor  43.718518   
4         M7A  Downtown Toronto                      Queen's Park  43.662301   
5         M9A         Etobicoke                  Islington Avenue  43.667856   
6         M1B       Scarborough                    Rouge, Malvern  43.806686   
7         M3B        North York                   Don Mills North  43.745906   
8         M4B         East York   Woodbine Gardens, Parkview Hill  43.706397   
9         M5B  Downtown Toronto          Ryerson, Garden District  43.657162   

   Longitude  
0 -79.329656  
1 -79.315572  
2 -79.360636  
3 -79.464763  
4 -79.389494  
5 -79.532242  
6 -79.194353  
7 -79.352188  
8 -79.309937  
9 -79.378937

# The end of the second part

In [19]:
CLIENT_ID = 'G1QYWBPLQJ2BJKYRNWAXU3QTLZAFOUW3VZO15OLXQ3IZHRBS' # your Foursquare ID
CLIENT_SECRET = 'NLWMY0GFVIUR1I3KUS2L41RE2DKRCSNC3R4JVCMAFPPUSP4P' # your Foursquare Secret
VERSION = '20180604'

print('Credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Credentails:
CLIENT_ID: G1QYWBPLQJ2BJKYRNWAXU3QTLZAFOUW3VZO15OLXQ3IZHRBS
CLIENT_SECRET:NLWMY0GFVIUR1I3KUS2L41RE2DKRCSNC3R4JVCMAFPPUSP4P


## 3. Explore and cluster the neighborhoods in Toronto

In [20]:
# define function to get venues
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [21]:
LIMIT = 100
toronto_venues = getNearbyVenues(names=toronto_data['Neighborhood'],
                                   latitudes=toronto_data['Latitude'],
                                   longitudes=toronto_data['Longitude']
                                  )

In [22]:
# check data
print(toronto_venues.shape)
toronto_venues.head()

(2251, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0         Parkwoods              43.753259              -79.329656   
1         Parkwoods              43.753259              -79.329656   
2         Parkwoods              43.753259              -79.329656   
3  Victoria Village              43.725882              -79.315572   
4  Victoria Village              43.725882              -79.315572   

                        Venue  Venue Latitude  Venue Longitude  \
0             Brookbanks Park       43.751976       -79.332140   
1               Variety Store       43.751974       -79.333114   
2  TTC stop - 44 Valley Woods       43.755402       -79.333741   
3      Victoria Village Arena       43.723481       -79.315635   
4                 Tim Hortons       43.725517       -79.313103   

      Venue Category  
0               Park  
1  Food & Drink Shop  
2           Bus Stop  
3       Hockey Arena  
4        Coffee Shop

## There is Venue Category - Neighborhood. Let's exclude these row

In [24]:
toronto_venues_en = toronto_venues[toronto_venues['Venue Category']!='Neighborhood']

In [25]:
len(toronto_venues_en['Neighborhood'].unique())

100

### Only 100 Neighborhoods out of 103 have some venues around

In [26]:
toronto_venues_en.shape

(2247, 7)

In [27]:
toronto_venues_en.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Adelaide, King, Richmond                                               99   
Agincourt                                                               5   
Agincourt North, L'Amoreaux East, Milliken, Ste...                      2   
Albion Gardens, Beaumond Heights, Humbergate, J...                     12   
Alderwood, Long Branch                                                 10   
Bathurst Manor, Downsview North, Wilson Heights                        20   
Bayview Village                                                         4   
Bedford Park, Lawrence Manor East                                      23   
Berczy Park                                                            57   
Birch Cliff, Cliffside West                                             4   
Bloordale Gardens, Eringate, Markland Wood, Old...                     10   
Brockton, Exhibition Place, Parkdale Village                           23   
Business Reply Mail Processing Centre 969 Eastern                      18   
CFB Toronto, Downsview East                                             3   
CN Tower, Bathurst Quay, Island airport, Harbou...                     17   
Cabbagetown, St. James Town                                            49   
Caledonia-Fairbanks                                                     4   
Canada Post Gateway Processing Centre                                  13   
Cedarbrae                                                               9   
Central Bay Street                                                     79   
Chinatown, Grange Park, Kensington Market                              87   
Christie                                                               18   
Church and Wellesley                                                   86   
Clairlea, Golden Mile, Oakridge                                         9   
Clarks Corners, Sullivan, Tam O'Shanter                                14   
Cliffcrest, Cliffside, Scarborough Village West                         2   
Cloverdale, Islington, Martin Grove, Princess G...                      1   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                             35   
Davisville North                                                        8   
Deer Park, Forest Hill SE, Rathnelly, South Hil...                     15   
Del Ray, Keelesdale, Mount Dennis, Silverthorn                          4   
Design Exchange, Toronto Dominion Centre                              100   
Don Mills North                                                         6   
Dorset Park, Scarborough Town Centre, Wexford H...                      8   
Dovercourt Village, Dufferin                                           18   
Downsview Central                                                       3   
Downsview Northwest                                                     4   
Downsview West                                                          4   
Downsview, North Park, Upwood Park                                      4   
East Birchmount Park, Ionview, Kennedy Park                             6   
East Toronto                                                            3   
Emery, Humberlea                                                        1   
Fairview, Henry Farm, Oriole                                           68   
First Canadian Place, Underground city                                100   
Flemingdon Park, Don Mills South                                       21   
Forest Hill North, Forest Hill West                                     5   
Glencairn                                                               4   
Guildwood, Morningside, West Hill                                       7   
Harbord, University of Toronto                                         35   
Harbourfront                                    

In [28]:
print('There are {} uniques categories.'.format(len(toronto_venues_en['Venue Category'].unique())))

There are 269 uniques categories.


## Analyze each neighborhood

In [29]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues_en[['Venue Category']], prefix="", prefix_sep="")


col = list(toronto_onehot.columns) # names of columns
# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# put neighborhood column to the first column
fixed_columns = ['Neighborhood'] + col
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head(10)

Neighborhood  Accessories Store  Afghan Restaurant  Airport  \
0         Parkwoods                  0                  0        0   
1         Parkwoods                  0                  0        0   
2         Parkwoods                  0                  0        0   
3  Victoria Village                  0                  0        0   
4  Victoria Village                  0                  0        0   
5  Victoria Village                  0                  0        0   
6  Victoria Village                  0                  0        0   
7  Victoria Village                  0                  0        0   
8      Harbourfront                  0                  0        0   
9      Harbourfront                  0                  0        0   

   Airport Food Court  Airport Lounge  Airport Service  Airport Terminal  \
0                   0               0                0                 0   
1                   0               0                0                 0   
2                   0               0                0                 0   
3                   0               0                0                 0   
4                   0               0                0                 0   
5                   0               0                0                 0   
6                   0               0                0                 0   
7                   0               0                0                 0   
8                   0               0                0                 0   
9                   0               0                0                 0   

   American Restaurant  Antique Shop  Aquarium  Art Gallery  \
0                    0             0         0            0   
1                    0             0         0            0   
2                    0             0         0            0   
3                    0             0         0            0   
4                    0             0         0            0   
5                    0             0         0            0   
6                    0             0         0            0   
7                    0             0         0            0   
8                    0             0         0            0   
9                    0             0         0            0   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  Auto Garage  \
0                    0                 0                   0            0   
1                    0                 0                   0            0   
2                    0                 0                   0            0   
3                    0                 0                   0            0   
4                    0                 0                   0            0   
5                    0                 0                   0            0   
6                    0                 0                   0            0   
7                    0                 0                   0            0   
8                    0                 0                   0            0   
9                    0                 0                   0            0   

   Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  \
0              0          0           0           0       0     0    0   
1              0          0           0           0       0     0    0   
2              0          0           0           0       0     0    0   
3              0          0           0           0       0     0    0   
4              0          0           0           0       0     0    0   
5              0          0           0           0       0     0    0   
6              0          0           0           0       0     0    0   
7              0          0           0           0       0     0    0   
8              0          0           0           0       1     0    0   
9              0          0           0           0       0     0    0   

   Baseball Field  Baseball Stadium  Baske

In [30]:
toronto_onehot.shape

(2247, 270)

### Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [31]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped.head(10)

Neighborhood  Accessories Store  \
0                           Adelaide, King, Richmond                0.0   
1                                          Agincourt                0.0   
2  Agincourt North, L'Amoreaux East, Milliken, St...                0.0   
3  Albion Gardens, Beaumond Heights, Humbergate, ...                0.0   
4                             Alderwood, Long Branch                0.0   
5    Bathurst Manor, Downsview North, Wilson Heights                0.0   
6                                    Bayview Village                0.0   
7                  Bedford Park, Lawrence Manor East                0.0   
8                                        Berczy Park                0.0   
9                        Birch Cliff, Cliffside West                0.0   

   Afghan Restaurant  Airport  Airport Food Court  Airport Lounge  \
0                0.0      0.0                 0.0             0.0   
1                0.0      0.0                 0.0             0.0   
2                0.0      0.0                 0.0             0.0   
3                0.0      0.0                 0.0             0.0   
4                0.0      0.0                 0.0             0.0   
5                0.0      0.0                 0.0             0.0   
6                0.0      0.0                 0.0             0.0   
7                0.0      0.0                 0.0             0.0   
8                0.0      0.0                 0.0             0.0   
9                0.0      0.0                 0.0             0.0   

   Airport Service  Airport Terminal  American Restaurant  Antique Shop  \
0              0.0               0.0             0.020202           0.0   
1              0.0               0.0             0.000000           0.0   
2              0.0               0.0             0.000000           0.0   
3              0.0               0.0             0.000000           0.0   
4              0.0               0.0             0.000000           0.0   
5              0.0               0.0             0.000000           0.0   
6              0.0               0.0             0.000000           0.0   
7              0.0               0.0             0.043478           0.0   
8              0.0               0.0             0.000000           0.0   
9              0.0               0.0             0.000000           0.0   

   Aquarium  Art Gallery  Arts & Crafts Store  Asian Restaurant  \
0       0.0     0.000000                  0.0          0.020202   
1       0.0     0.000000                  0.0          0.000000   
2       0.0     0.000000                  0.0          0.000000   
3       0.0     0.000000                  0.0          0.000000   
4       0.0     0.000000                  0.0          0.000000   
5       0.0     0.000000                  0.0          0.000000   
6       0.0     0.000000                  0.0          0.000000   
7       0.0     0.000000                  0.0          0.000000   
8       0.0     0.017544                  0.0          0.000000   
9       0.0     0.000000                  0.0          0.000000   

   Athletics & Sports  Auto Garage  Auto Workshop  BBQ Joint  Baby Store  \
0                 0.0          0.0            0.0   0.000000         0.0   
1                 0.0          0.0            0.0   0.000000         0.0   
2                 0.0          0.0            0.0   0.000000         0.0   
3                 0.0          0.0            0.0   0.000000         0.0   
4                 0.1          0.0            0.0   0.000000         0.0   
5                 0.0          0.0            0.0   0.000000         0.0   
6                 0.0          0.0            0.0   0.000000         0.0   
7                 0.0          0.0            0.0   0.000000         0.0   
8                 0.0          0.0            0.0   0.017544         0.0   
9                 0.0          0.0            0.0   0.000000         0.0   

   Bagel Shop    Bakery  Bank       Bar  Baseball Fie

In [32]:
toronto_grouped.shape

(100, 270)

## Let's put most common venues into a pandas dataframe

In [33]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [34]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood  \
0                           Adelaide, King, Richmond   
1                                          Agincourt   
2  Agincourt North, L'Amoreaux East, Milliken, St...   
3  Albion Gardens, Beaumond Heights, Humbergate, ...   
4                             Alderwood, Long Branch   

       1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0                Coffee Shop                   Bar            Restaurant   
1  Latin American Restaurant                Lounge          Skating Rink   
2                 Playground                  Park           Yoga Studio   
3              Grocery Store           Pizza Place            Beer Store   
4                Pizza Place           Coffee Shop          Skating Rink   

  4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
0                  Café       Thai Restaurant      Sushi Restaurant   
1        Breakfast Spot        Clothing Store           Yoga Studio   
2            Donut Shop    Dim Sum Restaurant                 Diner   
3   Fried Chicken Joint   Japanese Restaurant  Fast Food Restaurant   
4                   Gym              Pharmacy        Sandwich Place   

  7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0            Steakhouse                 Hotel          Concert Hall   
1             Drugstore   Distribution Center               Dog Run   
2        Discount Store   Distribution Center               Dog Run   
3          Liquor Store        Discount Store        Sandwich Place   
4                   Pub                  Pool    Athletics & Sports   

  10th Most Common Venue  
0                 Lounge  
1       Doner Restaurant  
2       Doner Restaurant  
3               Pharmacy  
4    Distribution Center

# Find Clusters

In [35]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([2, 2, 1, 2, 2, 2, 2, 2, 2, 2], dtype=int32)

In [42]:
kmeans.labels_

array([2, 2, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 1, 2, 2, 1, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 3, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 1, 0, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 0, 2, 2, 2, 2, 2, 2, 1, 2, 2,
       2, 2, 2, 2, 2, 2, 1, 2, 1, 4, 2, 2, 2, 1, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 1, 2, 2, 2, 2, 2, 1], dtype=int32)

In [36]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = toronto_data

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head() # check the last columns!

Postal Code           Borough                      Neighborhood   Latitude  \
0         M3A        North York                         Parkwoods  43.753259   
1         M4A        North York                  Victoria Village  43.725882   
2         M5A  Downtown Toronto                      Harbourfront  43.654260   
3         M6A        North York  Lawrence Heights, Lawrence Manor  43.718518   
4         M7A  Downtown Toronto                      Queen's Park  43.662301   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0 -79.329656             1.0              Bus Stop                  Park   
1 -79.315572             2.0          Hockey Arena          Intersection   
2 -79.360636             2.0           Coffee Shop                   Pub   
3 -79.464763             2.0     Accessories Store        Clothing Store   
4 -79.389494             2.0           Coffee Shop                  Park   

  3rd Most Common Venue       4th Most Common Venue  5th Most Common Venue  \
0     Food & Drink Shop                 Yoga Studio             Donut Shop   
1           Coffee Shop  Financial or Legal Service  Portuguese Restaurant   
2                Bakery                        Park                   Café   
3              Boutique                 Event Space            Coffee Shop   
4                   Gym                Burger Joint              Nightclub   

   6th Most Common Venue   7th Most Common Venue 8th Most Common Venue  \
0         Discount Store     Distribution Center               Dog Run   
1            Yoga Studio          Discount Store   Distribution Center   
2         Breakfast Spot              Restaurant               Theater   
3  Vietnamese Restaurant  Furniture / Home Store    Miscellaneous Shop   
4     Seafood Restaurant          Sandwich Place         Burrito Place   

  9th Most Common Venue 10th Most Common Venue  
0      Doner Restaurant              Drugstore  
1               Dog Run       Doner Restaurant  
2    Mexican Restaurant                  Hotel  
3      Doner Restaurant         Discount Store  
4                  Café  Portuguese Restaurant

In [47]:
toronto_merged.shape

(103, 16)

### drop Neighborhoods without venues, so without cluster labels

In [49]:
toronto_merged.dropna(axis = 0, inplace = True)

In [50]:
toronto_merged.shape

(100, 16)

In [51]:
address = 'Toronto, Canada'

geolocator = Nominatim(user_agent="can_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.653963, -79.387207.


## Visualization

In [52]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

# Examine clusters

## Cluster 1 (Cluster Label = 0)

In [54]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[2] + list(range(6, toronto_merged.shape[1]))]]

Neighborhood 1st Most Common Venue  \
57                                    Emery, Humberlea        Baseball Field   
101  Humber Bay, King's Mill Park, Kingsway Park So...        Baseball Field   

    2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
57            Yoga Studio             Drugstore        Discount Store   
101           Yoga Studio             Drugstore        Discount Store   

    5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
57    Distribution Center               Dog Run      Doner Restaurant   
101   Distribution Center               Dog Run      Doner Restaurant   

    8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
57             Donut Shop   Dumpling Restaurant   Fast Food Restaurant  
101            Donut Shop   Dumpling Restaurant   Fast Food Restaurant

## Cluster 2 (Cluster Label = 1)

In [55]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[2] + list(range(6, toronto_merged.shape[1]))]]

Neighborhood 1st Most Common Venue  \
0                                           Parkwoods              Bus Stop   
21                                Caledonia-Fairbanks                  Park   
32                                Scarborough Village            Playground   
35                                       East Toronto                  Park   
40                        CFB Toronto, Downsview East                  Park   
61                                      Lawrence Park                  Park   
64                                             Weston                  Park   
66                                    York Mills West                  Park   
85  Agincourt North, L'Amoreaux East, Milliken, St...            Playground   
91                                           Rosedale                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                   Park     Food & Drink Shop           Yoga Studio   
21         Women's Store                Market           Yoga Studio   
32     Convenience Store           Yoga Studio             Drugstore   
35           Coffee Shop     Convenience Store           Yoga Studio   
40               Airport           Snack Place           Yoga Studio   
61              Bus Line           Swim School           Yoga Studio   
64     Convenience Store           Yoga Studio             Drugstore   
66                  Bank     Convenience Store           Yoga Studio   
85                  Park           Yoga Studio            Donut Shop   
91                 Trail            Playground           Yoga Studio   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0             Donut Shop        Discount Store   Distribution Center   
21            Donut Shop                 Diner        Discount Store   
32                 Diner        Discount Store   Distribution Center   
35             Drugstore        Discount Store   Distribution Center   
40            Donut Shop                 Diner        Discount Store   
61            Donut Shop        Discount Store   Distribution Center   
64        Discount Store   Distribution Center               Dog Run   
66             Drugstore        Discount Store   Distribution Center   
85    Dim Sum Restaurant                 Diner        Discount Store   
91    Dim Sum Restaurant                 Diner        Discount Store   

   8th Most Common Venue 9th Most Common Venue       10th Most Common Venue  
0                Dog Run      Doner Restaurant                    Drugstore  
21   Distribution Center               Dog Run             Doner Restaurant  
32               Dog Run      Doner Restaurant                   Donut Shop  
35               Dog Run      Doner Restaurant                   Donut Shop  
40   Distribution Center               Dog Run             Doner Restaurant  
61               Dog Run      Doner Restaurant                    Drugstore  
64      Doner Restaurant            Donut Shop  Eastern European Restaurant  
66               Dog Run      Doner Restaurant                   Donut Shop  
85   Distribution Center               Dog Run             Doner Restaurant  
91   Distribution Center               Dog Run             Doner Restaurant

## Cluster 3 (Cluster Label = 2)

In [56]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[2] + list(range(6, toronto_merged.shape[1]))]]

Neighborhood  \
1                                     Victoria Village   
2                                         Harbourfront   
3                     Lawrence Heights, Lawrence Manor   
4                                         Queen's Park   
6                                       Rouge, Malvern   
7                                      Don Mills North   
8                      Woodbine Gardens, Parkview Hill   
9                             Ryerson, Garden District   
10                                           Glencairn   
12              Highland Creek, Rouge Hill, Port Union   
13                    Flemingdon Park, Don Mills South   
14                                    Woodbine Heights   
15                                      St. James Town   
16                                  Humewood-Cedarvale   
17   Bloordale Gardens, Eringate, Markland Wood, Ol...   
18                   Guildwood, Morningside, West Hill   
19                                         The Beaches   
20                                         Berczy Park   
22                                              Woburn   
23                                             Leaside   
24                                  Central Bay Street   
25                                            Christie   
26                                           Cedarbrae   
27                                   Hillcrest Village   
28     Bathurst Manor, Downsview North, Wilson Heights   
29                                    Thorncliffe Park   
30                            Adelaide, King, Richmond   
31                        Dovercourt Village, Dufferin   
33                        Fairview, Henry Farm, Oriole   
34                     Northwood Park, York University   
36   Harbourfront East, Toronto Islands, Union Station   
37                            Little Portugal, Trinity   
38         East Birchmount Park, Ionview, Kennedy Park   
39                                     Bayview Village   
41                        The Danforth West, Riverdale   
42            Design Exchange, Toronto Dominion Centre   
43        Brockton, Exhibition Place, Parkdale Village   
44                     Clairlea, Golden Mile, Oakridge   
45                            Silver Hills, York Mills   
46                                      Downsview West   
47                      The Beaches West, India Bazaar   
48                      Commerce Court, Victoria Hotel   
49                  Downsview, North Park, Upwood Park   
50                                       Humber Summit   
51     Cliffcrest, Cliffside, Scarborough Village West   
53                                   Downsview Central   
54                                     Studio District   
55                   Bedford Park, Lawrence Manor East   
56      Del Ray, Keelesdale, Mount Dennis, Silverthorn   
58                         Birch Cliff, Cliffside West   
59                                    Willowdale South   
60                                 Downsview Northwest   
63                       The Junction North, Runnymede   
65   Dorset Park, Scarborough Town Centre, Wexford ...   
67                                    Davisville North   
68                 Forest Hill North, Forest Hill West   
69                       High Park, The Junction South   
70                                           Westmount   
71                                   Maryvale, Wexford   
72                                     Willowdale West   
73                                  North Toronto West   
74                 The Annex, North Midtown, Yorkville   
75                              Parkdale, Roncesvalles   
76               Canada Post Gateway Processing Centre   
77   Kingsview Village, Martin Grove Gardens, Richv...   
78                                           Agincourt   
79                                          Davisville   
80                      Harbord, University of Toronto   
81                                  Runn

## Cluster 4 (Cluster Label = 3)

In [57]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[2] + list(range(6, toronto_merged.shape[1]))]]

Neighborhood 1st Most Common Venue  \
11  Cloverdale, Islington, Martin Grove, Princess ...         Jewelry Store   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
11           Yoga Studio   Dumpling Restaurant        Discount Store   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
11   Distribution Center               Dog Run      Doner Restaurant   

   8th Most Common Venue 9th Most Common Venue       10th Most Common Venue  
11            Donut Shop             Drugstore  Eastern European Restaurant

## Cluster 5 (Cluster Label = 4)

In [58]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[2] + list(range(6, toronto_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
62     Roselawn                Garden           Yoga Studio   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
62            Donut Shop                 Diner        Discount Store   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
62   Distribution Center               Dog Run      Doner Restaurant   

   9th Most Common Venue 10th Most Common Venue  
62             Drugstore           Dessert Shop